### PR TITLE
Ability to pass custom headers for HTTP requests

### PR DESCRIPTION
**Pull request by av-ast**
I added ability to pass custom headers for HTTP requests.

### CODE REVIEW

**Comment by brian-brazil** [comment]
Why do you want this? The pushgateway doesn't use such headers.

**Comment by av-ast** [comment]
In my case I want to pass headers for Basic Auth.

File: spec/prometheus/client/push_spec.rb [hunk 1]
```diff
@@ -21,6 +21,22 @@
...
 
+    it 'allows to specify a custom Pushgateway with custom headers' do
+      headers = { 'Key' => 'Value' }
+      push = Prometheus::Client::Push.new('test-job', nil, nil, headers)
+
+      merged_headers = described_class::DEFAULT_HEADERS.merge(headers)
+      expect(push.headers).to eql(merged_headers)
+    end
+
+    it 'allows to averride default headers' do
```
Comment (grobie) [inline spec/prometheus/client/push_spec.rb]: What is the use case for setting a different content-type?

Comment (av-ast) [inline spec/prometheus/client/push_spec.rb]: Actually, there is no real reason to use a different content-type.
I just showed such ability with this test case...

**Comment by grobie** [comment]
I think this change is inline with our principle that clients should support standard authorization when talking to other servers, like our support for scrapes and in federation. Our standard answer to requests for https/etc. for pushgateway and other components is to use a reverse proxy in front of it, therefore clients should have the ability to set required headers. Any concerns with that @brian-brazil?

I'm still not really sure why we should allow to change the Content-Type, but on the other hand, I don't see a strong reason to forbid it.

**Comment by brian-brazil** [comment]
I'd suggest having explicit basic auth support, rather than requiring the user to hand-assemble the Authentication header.

In Python there's a handler you can use for this: https://github.com/prometheus/client_python/#handlers-for-authentication
In Java it's not released yet, but user/password in the passed URL object work. There may be further extensions to support SSL and bearer tokens.

> I'm still not really sure why we should allow to change the Content-Type, but on the other hand, I don't see a strong reason to forbid it.

It's allow the user to set an incorrect content type, the library is choosing the content so it alone should specify the type.

**Comment by grobie** [comment]
Resolved by #74.

**Pull request closed without merging**